### PR TITLE
gNMI Subscribe Client Setup

### PR DESCRIPTION
**Pull request by Jared-Mullin**
Re: Issue #192 
This PR sets up flags, path parsing and constructs the initial SubscribeRequest

### CODE REVIEW

File: gnmi_subscribe/gnmi_subscribe.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	"github.com/google/gnxi/utils/credentials"
+	"github.com/google/gnxi/utils/xpath"
+	pb "github.com/openconfig/gnmi/proto/gnmi"
+	"google.golang.org/grpc"
+)
+
+type arrayFlags []string
+
+func (i *arrayFlags) String() string {
+	return "my string representation"
```
Comment (ericm) [inline gnmi_subscribe/gnmi_subscribe.go:33]: Is this correct? If it just requires you to implement an interface maybe just return `""`.

Comment (Jared-Mullin) [reply gnmi_subscribe/gnmi_subscribe.go:33]: Agreed

File: gnmi_subscribe/gnmi_subscribe.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+)
+
+var (
+	subscribeClient    pb.GNMI_SubscribeClient
+	xPathFlags         arrayFlags
+	pbPathFlags        arrayFlags
+	targetAddr         = flag.String("target_addr", ":9339", "The target address in the format of host:port")
+	targetName         = flag.String("target_name", "", "The target name used to verify the hostname returned by TLS handshake")
+	connectionTimeout  = flag.Duration("timeout", defaultTimeout, "The timeout for a request, 10 seconds by default")
+	valueStreamingMode = flag.String("value_mode", "STREAM", "The mode of streaming values the target should use use, STREAM by default")
```
Comment (samribeiro) [inline gnmi_subscribe/gnmi_subscribe.go:52]: I would define instead two booleans.
-once
-pool
 If none is mentioned then use stream (default). If both are defined, send error that only one of either can be used.

Comment (Jared-Mullin) [reply gnmi_subscribe/gnmi_subscribe.go:52]: That's a lot cleaner, thank you

File: gnmi_subscribe/gnmi_subscribe.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	subscriptionMode   = flag.String("subscription_mode", "TARGET_DEFINED", "The subscription mode the target should use, TARGET_DEFINED by default")
```
Comment (samribeiro) [inline gnmi_subscribe/gnmi_subscribe.go:53]: same, if -stream_on_change is defined then use ON_CHANGE. If not defined, use TARGET_DEFINED.
If -sample_interval is defined use SAMPLE.

File: gnmi_subscribe/gnmi_subscribe.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	encodingFormat     = flag.String("encoding", "JSON_IETF", "The encoding format used by the target for notifications")
+	allowAggregation   = flag.Bool("allow_aggregation", false, "If true, the elements which are marked eligible are aggregated")
```
Comment (samribeiro) [inline gnmi_subscribe/gnmi_subscribe.go:55]: no need to support this.

File: gnmi_subscribe/gnmi_subscribe.go (hunk 3 above)
Comment (samribeiro) [reply gnmi_subscribe/gnmi_subscribe.go:53]: also add:
 bool suppress_redundant (default false)
 uint64 heartbeat_interval ((default 10x sample_interval)
This is used if SAMPLE mode is used.

File: gnmi_subscribe/gnmi_subscribe.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	var heartbeat uint64
+	switch {
+	case *streamOnChange && *sampleInterval != 0:
+		return nil, errors.New("Only one of -stream_on_change and -sample_interval can be set")
+	case *streamOnChange:
+		subscriptionMode = pb.SubscriptionMode_ON_CHANGE
+	case *sampleInterval != 0:
+		subscriptionMode = pb.SubscriptionMode_SAMPLE
+		if *heartbeatInterval == 0 {
+			heartbeat = *sampleInterval * 10
```
Comment (samribeiro) [inline gnmi_subscribe/gnmi_subscribe.go:174]: After thinking about it, I take it back, lets not enforce a default. Use it only if set.

Comment (Jared-Mullin) [reply gnmi_subscribe/gnmi_subscribe.go:174]: Sounds good!

File: gnmi_subscribe/gnmi_subscribe.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/google/gnxi/utils/credentials"
+	"github.com/google/gnxi/utils/xpath"
+	"github.com/openconfig/gnmi/proto/gnmi"
+	pb "github.com/openconfig/gnmi/proto/gnmi"
+	"google.golang.org/grpc"
+)
+
+type arrayFlags []string
+
+func (i *arrayFlags) String() string {
```
Comment (brianokelly1) [inline gnmi_subscribe/gnmi_subscribe.go:34]: I might be a touch unfamiliar with the code base (I see this approach also used in gnmi_set), it's not clear to me why you wouldn't just set xPathFlags and pbPathFlags as type []string and not use this new type.

We don't seem to use either the Set or String functionality (in gnmi_set either). Could you provide some background here (apologies if you are having to explain the codebase to me!).

File: gnmi_subscribe/gnmi_subscribe.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	return ""
+}
+
+func (i *arrayFlags) Set(value string) error {
+	*i = append(*i, value)
+	return nil
+}
+
+const (
+	defaultTimeout = 10
```
Comment (brianokelly1) [inline gnmi_subscribe/gnmi_subscribe.go:44]: Add a comment, what is it the default timeout for. Also include the unit in the description as for what 10 represents (seconds, minutes).

Comment (Jared-Mullin) [reply gnmi_subscribe/gnmi_subscribe.go:44]: Agreed, thank you!

**Pull request merged**